### PR TITLE
[prometheus] Improove Prometheus Retain alerts

### PR DESCRIPTION
**Pull request by vitaliy-sn**
Signed-off-by: Vitaliy Snurnitsin <vitaliy.snurnitsin@flant.com>

## Description

Prometheus Retain alerts now respect the retentionDays parameters.

## Why do we need it, and what problem does it solve?

Ignoring the retentionDays parameters leads to false alerts.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: prometheus
type: fix
summary: Improve Prometheus Retain alerts.
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/gcp'...

File: modules/300-prometheus/hooks/metrics_storage_retention.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2021 Flant JSC
```
Comment (apolovov) [inline modules/300-prometheus/hooks/metrics_storage_retention.go:2]:
```suggestion
Copyright 2022 Flant JSC
```

File: modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml [hunk 2]
```diff
@@ -18,11 +18,11 @@
         ```
         Consider increasing it https://deckhouse.io/en/documentation/v1/modules/300-prometheus/faq.html#how-to-expand-disk-size
   # If there was retention in the last 24 hours and the oldest metric was recorded less than 10 days ago and storage.tsdb.retention.size has not been changed.
-  - alert: PrometheusMainRetainLessThan10Days
+  - alert: PrometheusMainRetainLessThanRequiredDays
     expr: |
       group by (service) (increase(prometheus_tsdb_size_retentions_total{service="prometheus"}[24h]) > 0)
       and
-      group by (service) ((time() - prometheus_tsdb_lowest_timestamp_seconds{service="prometheus"}) / 60 / 60 / 24 < 10)
+      group by (service) ((time() - prometheus_tsdb_lowest_timestamp_seconds{service="prometheus"}) / 60 / 60 / 24 < d8_prometheus_main_storage_retention_days)
```
Comment (apolovov) [inline modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml:25]:
```suggestion
      group by (service) ((time() - prometheus_tsdb_lowest_timestamp_seconds{service="prometheus"}) / 60 / 60 / 24 < d8_prometheus_storage_retention_days{"prometheus":"main"})
```

File: modules/300-prometheus/hooks/metrics_storage_retention.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	Queue:        "/modules/prometheus/metrics_storage_retention",
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10},
+}, storageRetentionMetricHandler)
+
+func storageRetentionMetricHandler(input *go_hook.HookInput) error {
+	retentionDaysMain := input.Values.Get("prometheus.retentionDays")
+	retentionDaysLongterm := input.Values.Get("prometheus.longtermRetentionDays")
+
+	input.MetricsCollector.Set(
```
Comment (apolovov) [inline modules/300-prometheus/hooks/metrics_storage_retention.go:34]: Let it be a single metric `d8_prometheus_storage_retention_days` with labels prometheus="main|longrerm"

File: modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml [hunk 4]
```diff
@@ -49,6 +49,6 @@
       plk_markup_format: markdown
       plk_protocol_version: "1"
       for: "10m"
-      summary: Prometheus-longterm data is retained for less than 30 days
+      summary: Prometheus-longterm data is retained for less than retention days
```
Comment (apolovov) [inline modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml:52]:
```suggestion
      summary: Prometheus-longterm data is being rotated earlier than configured retention days
```

File: modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml [hunk 5]
```diff
@@ -49,6 +49,6 @@
...
       description: |
         You need to increase the disk size or reduce the number of metrics
```
Comment (apolovov) [inline modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml:54]:
```suggestion
        You need to increase the disk size, reduce the number of metrics or decrease `longtermRetentionDays` module parameter.
```

**Review by apolovov — CHANGES_REQUESTED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/gcp'...

File: modules/300-prometheus/hooks/metrics_storage_retention.go (hunk 1 above)
Comment (vitaliy-sn) [reply modules/300-prometheus/hooks/metrics_storage_retention.go]: Fixed

File: modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml (hunk 2 above)
Comment (vitaliy-sn) [reply modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml]: Fixed

File: modules/300-prometheus/hooks/metrics_storage_retention.go (hunk 3 above)
Comment (vitaliy-sn) [reply modules/300-prometheus/hooks/metrics_storage_retention.go:34]: Fixed

File: modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml (hunk 4 above)
Comment (vitaliy-sn) [reply modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml]: Fixed

File: modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml (hunk 5 above)
Comment (vitaliy-sn) [reply modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml]: Fixed

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/gcp'...

File: modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml [hunk 6]
```diff
@@ -18,11 +18,11 @@
         ```
         Consider increasing it https://deckhouse.io/en/documentation/v1/modules/300-prometheus/faq.html#how-to-expand-disk-size
   # If there was retention in the last 24 hours and the oldest metric was recorded less than 10 days ago and storage.tsdb.retention.size has not been changed.
-  - alert: PrometheusMainRetainLessThan10Days
+  - alert: PrometheusMainRetainLessThanRequiredDays
```
Comment (apolovov) [inline modules/300-prometheus/monitoring/prometheus-rules/prometheus-storage.yaml:21]: Required?
Configured

**Review by apolovov — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Cherry pick PR [2892](https://github.com/deckhouse/deckhouse/pull/2892) to the branch [release-1.38](https://github.com/deckhouse/deckhouse/tree/release-1.38) successful!

**Pull request merged**